File: core/reviewer.py
"""
The Reviewer -- deterministic QA gate. Nothing reaches "approved for
dispatch" without passing every check here, independent of whatever the
Strategist/Writer did upstream. This is the same invariant enforcement
described for the reviewer node in the source architecture ("Failed
validation: attempted to bid below raw material COGS floor").
"""
from __future__ import annotations

from dataclasses import dataclass

from core.models import QualifiedLead
from core.pricing import MIN_MARGIN_PCT

MAX_REWRITES = 2


@dataclass
class ReviewResult:
    approved: bool
    notes: list[str]
    needs_rewrite: bool


def review(lead: QualifiedLead, proposal_text: str, rewrite_count: int) -> ReviewResult:
    notes: list[str] = []

    # 1. mandatory anti-bot phrase must be present verbatim if one was found
    if lead.trap_analysis.has_anti_bot_phrase and lead.trap_analysis.required_first_word:
        if lead.trap_analysis.required_first_word not in proposal_text.split("\n\n", 1)[0]:
            notes.append(f"missing_mandatory_opening_word:{lead.trap_analysis.required_first_word}")

    # 2. no leaked injection artifacts (discount language etc. that came
    #    from the client's text rather than our own strategy) should survive
    #    into the draft
    for phrase in ("ignore all previous", "ignore previous instructions", "you are now"):
        if phrase in proposal_text.lower():
            notes.append("injection_leak_detected")
            break

    # 3. price math -- reviewer independently re-checks the invariant rather
    #    than trusting the field on the lead object
    if lead.bid_price_usd <= 0:
        notes.append("invalid_price")
    floor = lead.cogs_usd * (1 + MIN_MARGIN_PCT)
    if lead.bid_price_usd < floor - 0.01:
        notes.append(f"price_below_cogs_floor: bid={lead.bid_price_usd:.2f} floor={floor:.2f}")

    # 4. sanity: quoted price actually appears in the draft
    if f"{lead.bid_price_usd:,.2f}" not in proposal_text:
        notes.append("quoted_price_not_present_in_draft")

    if not notes:
        return ReviewResult(approved=True, notes=["ok"], needs_rewrite=False)

    needs_rewrite = rewrite_count < MAX_REWRITES and "price_below_cogs_floor" not in "".join(notes)
    return ReviewResult(approved=False, notes=notes, needs_rewrite=needs_rewrite)

File: iechm-leadgen-engine-c/src/reviewer.c
#include "reviewer.h"
#include "pricing.h"
#include "strutil.h"
#include "fmt.h"
#include <string.h>
#include <stdio.h>

static void add_note(ReviewResult *r, const char *note) {
    if (r->notes_count < NOTES_MAX) xcpy(r->notes[r->notes_count++], NOTE_LEN, note);
}

ReviewResult review_proposal(const QualifiedLead *lead, const char *proposal_text, int rewrite_count) {
    ReviewResult r;
    memset(&r, 0, sizeof r);

    if (lead->trap_analysis.has_anti_bot_phrase && lead->trap_analysis.has_required_first_word) {
        const char *sep = strstr(proposal_text, "\n\n");
        char first_para[TEXT_LEN];
        if (sep) slice_copy(first_para, sizeof first_para, proposal_text, 0, (int) (sep - proposal_text));
        else xcpy(first_para, sizeof first_para, proposal_text);
        if (strstr(first_para, lead->trap_analysis.required_first_word) == NULL) {
            char note[NOTE_LEN];
            snprintf(note, sizeof note, "missing_mandatory_opening_word:%s", lead->trap_analysis.required_first_word);
            add_note(&r, note);
        }
    }

    if (ci_find(proposal_text, "ignore all previous", 0) >= 0 ||
        ci_find(proposal_text, "ignore previous instructions", 0) >= 0 ||
        ci_find(proposal_text, "you are now", 0) >= 0) {
        add_note(&r, "injection_leak_detected");
    }

    bool has_floor_note = false;
    if (lead->bid_price_usd <= 0.0) add_note(&r, "invalid_price");
    double floor = lead->cogs_usd * (1.0 + MIN_MARGIN_PCT);
    if (lead->bid_price_usd < floor - 0.01) {
        char note[NOTE_LEN];
        snprintf(note, sizeof note, "price_below_cogs_floor: bid=%.2f floor=%.2f", lead->bid_price_usd, floor);
        add_note(&r, note);
        has_floor_note = true;
    }

    char price_str[48];
    format_money_commas(lead->bid_price_usd, price_str, sizeof price_str);
    if (strstr(proposal_text, price_str) == NULL) add_note(&r, "quoted_price_not_present_in_draft");

    if (r.notes_count == 0) {
        r.approved = true;
        add_note(&r, "ok");
        r.needs_rewrite = false;
        return r;
    }

    r.approved = false;
    r.needs_rewrite = (rewrite_count < MAX_REWRITES) && !has_floor_note;
    return r;
}
